nocap_herald.c

```diff
@@ -0,0 +1,189 @@
+
+	// Set io vector to the netdump message buffer
+	iov.iov_base = &ndmsg;
+	iov.iov_len = sizeof(ndmsg);
+
+	// Track the socket and message together in msg
+	msg.msg_name = &ss;
+	msg.msg_namelen = sizeof(ss);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	// Set the size in bytes for control messages and track size in cmsgcz
+	cmsgsz = CMSG_SPACE(sizeof(struct in_addr));
+
+	// Allocate memory for the control messages and store size
+	msg.msg_control = calloc(1, cmsgsz);
+	if (msg.msg_control == NULL) {
+		error = errno;
+		goto out;
+	}
+	msg.msg_controllen = cmsgsz;
+
+	// Read message from server socket in to msg, should include path
+	len = recvmsg(g_sock, &msg, 0);
+	if (len < 0) {
+		error = errno;
+		goto out;
+	}
+
+	// Confirm we got the entire message header
+	if ((size_t)len < sizeof(struct netdump_msg_hdr)) {
+		error = EINVAL;
+		goto out;
+	}
+
+	// Convert netdump header to standard header
+	ndtoh(&ndmsg.hdr);
+
+	// Error if message header type != HERALD
+	//  or original netdump header msg length - struct's length
+	//  or socket protocol != AF_INET
+	if (ndmsg.hdr.mh_type != NETDUMP_HERALD ||
+	    (size_t)len - sizeof(struct netdump_msg_hdr) != ndmsg.hdr.mh_len ||
+	    ss.ss_family != AF_INET) {
+		error = EINVAL;
+		goto out;
+	}
+
+	// Read control message header
+	cmh = CMSG_FIRSTHDR(&msg);
+	if (cmh->cmsg_level != IPPROTO_IP || cmh->cmsg_type != IP_RECVDSTADDR) {
+		error = EINVAL;
+		goto out;
+	}
+
+	// This is the IP receiving the HERALD msg
+	// Useful when netdump server is multi homed
+	dip = (struct in_addr *)(void *)CMSG_DATA(msg.msg_control);
+
+	// Setup new client socket
+	*nsd = socket(PF_INET, SOCK_DGRAM | SOCK_CLOEXEC | SOCK_NONBLOCK,
+	    IPPROTO_UDP);
+	if (*nsd < 0) {
+		error = errno;
+		goto out;
+	}
+
+	// Setup sockaddr_in for new listener on server
+	// Zero out socket in, copy contents from msg, set dip IP
+	memset(&sin, 0, sizeof(sin));
+	sin.sin_len = sizeof(sin);
+	sin.sin_family = AF_INET;
+	sin.sin_addr.s_addr = dip->s_addr;
+	sin.sin_port = htons(0);  // pick random port
+
+	if (bind(*nsd, (struct sockaddr *)&sin, sin.sin_len) != 0) {
+		error = errno;
+		goto out;
+	}
+
+	// Resuse sockaddr in msg for the ACK message
+	from = (struct sockaddr_in *)msg.msg_name;
+	from->sin_port = htons(NETDUMP_ACKPORT);
+	if (connect(*nsd, (struct sockaddr *)from, from->sin_len) != 0) {
+		error = errno;
+		goto out;
+	}
+
+	// Setup sockaddr to return for alloc_client later
+	memset(sinp, 0, sizeof(*sinp));
+	memcpy(sinp, msg.msg_name, msg.msg_namelen);
+
+	// Get the file path from the control message
+	pathsz = ndmsg.hdr.mh_len;
+
+	*pathp = NULL;
+	if (pathsz > 0 && pathsz <= MIN(MAXPATHLEN, NETDUMP_DATASIZE) &&
+	    ndmsg.data[pathsz - 1] == '\0')
+	    *pathp = strdup(ndmsg.data);
+
+	*seqno = ndmsg.hdr.mh_seqno;
+
+out:
+	// ?? Is this all I need to clean up?
+	if (msg.msg_control != NULL)
+		free(msg.msg_control);
+	if (error != 0 && *nsd >= 0)
+		(void)close(*nsd);
+	return (error);
+}
```

Comment:
It should be possible to avoid duplication with cap_herald.c by refactoring the code a bit:
- change netdump_cap_herald() to call herald_command() directly when capsicum is not configured,
- refactor herald_command() to avoid using nvlists directly, instead just using C out-parameters or a structure or something.

I believe you could also use libnv on Linux: https://github.com/fudosecurity/nvlist